### PR TITLE
[perf] Add configurable diffusion compile granularity

### PR DESCRIPTION
**Pull request by TaffyOfficial**
## Purpose

Add explicit diffusion compile controls through the supported CLI and per-stage configuration paths. The generic diffusion model runner now supports:

- `diffusion_compile_granularity: regional | full` (default: `regional`)
- `diffusion_compile_dynamic: bool` (default: `true`)

Equivalent CLI flags are `--diffusion-compile-granularity` and `--[no-]diffusion-compile-dynamic`. This replaces the environment-variable interface from the earlier revision.

The recommended fixed-shape LTX2.3 path remains regional compilation with `dynamic=false`. Full-transformer scope is experimental and is rejected before model loading when HSDP, sequence parallelism, CPU offload, or layerwise offload is enabled.

Compilation is lazy. Synchronous regional wrapper setup is now atomic and can safely continue with the uncompiled model if setup fails, while backend or graph errors may still surface on the first request. Pipelines with a custom `setup_compile()` implementation continue to own their compile policy independently.

## Performance evidence

The measurements below were captured on pre-interface-migration commit `5c90c055f9c62db241719de16d2fb051a15a658b`. The optimized runtime operation is unchanged: regional `torch.compile(..., dynamic=False)`. This revision changes how that setting is configured and adds validation/failure-safety.

No-profiler LTX2.3 T2V graph-mode benchmark (`/v1/videos/sync`, 512x384, 25 frames, 24 fps, 20 steps, concurrency 1, one tiny plus one full warmup discarded, three measured requests):

| Mode | Mean e2e | QPS | stage_0_gen_ms | Peak memory |
| --- | ---: | ---: | ---: | ---: |
| regional, dynamic=true | 4.3686s | 0.2289 | 4202.1 | 73290 MB |
| regional, dynamic=false | 3.8079s | 0.2626 | 3654.6 | 73290 MB |

Regional `dynamic=false` changed mean e2e by -12.83%, QPS by +14.72%, and stage generation time by -13.03%, with unchanged peak memory.

Paired profiler traces attribute the improvement to lower fixed-shape CPU-side overhead rather than different math kernels: trace events fell 12.0%, `cpu_op` events 28.1%, `aten::empty` 44.5%, and `aten::to` 63.3%; GEMM, FA3, RMSNorm, and CUDA launch counts were effectively unchanged.

The fixed-output comparison against the diffusers baseline passed the existing thresholds:

```text
25 frames compared
SSIM avg: 0.9644
SSIM min: 0.9618
PSNR avg: 36.38
PSNR min: 35.83
passes_thresholds: true
```

## Test plan

- CLI/default-stage propagation for both controls
- Deploy YAML, structured-config validation, and CLI precedence
- Runtime defaults, strict type/enum validation, and full-mode HSDP/SP/offload guards
- Regional/full runner dispatch and exact `dynamic` forwarding
- Atomic regional setup and truthful synchronous fallback behavior
- Ruff check and format validation
- Python bytecode compilation and `git diff --check`
- PR scope gate and DCO validation

The targeted pytest files cannot be collected in the local Windows checkout because its Python environment does not contain the `vllm` dependency. GitHub CI is the executable unit-test source for this revision.

### CODE REVIEW

**Comment by hsliuustc0106** [comment]
add profiling comparison

**Comment by TaffyOfficial** [comment]
> add profiling comparison

add

**Comment by zhumingjue138** [comment]
please add necessary UT cases

**Comment by TaffyOfficial** [comment]
Added lightweight UT coverage for the new compile controls.

Covered cases:
- default env preserves `regional dynamic=True`
- invalid env values fall back to defaults
- `VLLM_OMNI_DIFFUSION_REGIONAL_COMPILE_DYNAMIC=0` dispatches `regionally_compile(..., dynamic=False)`
- `VLLM_OMNI_DIFFUSION_COMPILE_GRANULARITY=full` dispatches `model.compile(dynamic=False)`

Verified with the targeted pytest command listed in the PR body (`4 passed`).

**Comment by mglyn** [comment]
A more fine-grained LTX2.3 block fusion change here #4790.
It touches a related compile-region issue around the masked FlashAttention prompt cross-attention path.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 from __future__ import annotations
 
 import copy
+import os
```
Comment (wtomin) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py]: There is no specific document for regional compilation. Could you write a doc specific to regional compilation feature, and also introduce the two env variables in this doc?

**Comment by wtomin** [comment]
How is the profiling result under high concurrency and mixed resolutions?

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 2]
```diff
@@ -46,6 +47,51 @@
 
 logger = init_logger(__name__)
 
+_REGIONAL_COMPILE_DYNAMIC_ENV = "VLLM_OMNI_DIFFUSION_REGIONAL_COMPILE_DYNAMIC"
```
Comment (lishunyang12) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py]: I think this name is going to trip people up now: `REGIONAL_COMPILE_DYNAMIC` also drives `dynamic` on the `full` path below, so someone running granularity=full won't expect a var with `REGIONAL` in it to matter. I'd drop `REGIONAL` (e.g. `..._COMPILE_DYNAMIC`) so the knob matches what it actually controls. 🤔

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 3]
```diff
@@ -142,8 +188,21 @@ def _compile_transformer(self, attr_name: str) -> None:
             return
 
         try:
-            setattr(self.pipeline, attr_name, regionally_compile(model, dynamic=True))
+            compile_dynamic = _resolve_regional_compile_dynamic()
+            compile_granularity = _resolve_compile_granularity()
+            if compile_granularity == "full":
```
Comment (lishunyang12) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py:157]: I'm not sure full-graph `model.compile()` composes with HSDP/USP wrapping the way `regionally_compile` does. My hunch is regional was picked partly so compile plays nice with the sharded/parallel transformer. If someone sets granularity=full while parallelism is on, could this silently corrupt the compiled regions rather than error? I'd add a guard (or at least a warning) that falls back to regional when `use_hsdp`/SP is active.

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 4]
```diff
@@ -142,8 +188,21 @@ def _compile_transformer(self, attr_name: str) -> None:
...
+                model.compile(dynamic=compile_dynamic)
```
Comment (lishunyang12) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py:158]: One thing I keep coming back to: the try/except promises an eager fallback, but `model.compile()` is lazy — it won't raise here, only on the first forward. So a bad `full` + `dynamic` combo skips this `except` entirely and blows up mid-generation instead of dropping to eager. I'd either note that in the log or trigger a trial forward so the fallback is real. 🙏

**Comment by TaffyOfficial** [comment]
> How is the profiling result under high concurrency and mixed resolutions?

Thanks. I reran the profiling on vLLM 0.25.0 with regional compilation and dynamic shapes enabled. The workload used LTX-2.3 T2V, 25 frames, 20 steps, 8 requests at concurrency 4, alternating between 512×384 and 768×512. All 8 requests succeeded in 38.09s, with 0.210 req/s throughput, 15.79s mean latency, 18.04s P50, 20.05s P95, and 72,922 MB reported peak memory.
One important observation is that StageRequestStats remained batch_size=1 for every request, even with max_num_seqs=4. An identical-prompt control produced the same behavior and nearly identical metrics. Therefore, these numbers represent queued high-concurrency mixed-resolution traffic rather than fused request batching. No OOMs or runtime compilation errors were observed.
I also added a dedicated regional compilation document covering both environment variables, their defaults, mixed-resolution guidance, lazy compilation behavior, and the full-compile parallelism fallback.

**Comment by TaffyOfficial** [comment]
@lishunyang12 fix

File: vllm_omni/diffusion/worker/diffusion_model_runner.py (hunk 1 above)
Comment (TaffyOfficial) [reply vllm_omni/diffusion/worker/diffusion_model_runner.py]: add

File: vllm_omni/diffusion/worker/diffusion_model_runner.py [hunk 5]
```diff
@@ -54,6 +55,51 @@
...
 
+_REGIONAL_COMPILE_DYNAMIC_ENV = "VLLM_OMNI_DIFFUSION_REGIONAL_COMPILE_DYNAMIC"
+_COMPILE_GRANULARITY_ENV = "VLLM_OMNI_DIFFUSION_COMPILE_GRANULARITY"
+_TRUE_ENV_VALUES = {"1", "true", "yes", "on"}
+_FALSE_ENV_VALUES = {"0", "false", "no", "off"}
+_COMPILE_GRANULARITIES = {"regional", "full"}
+
+
+def _resolve_regional_compile_dynamic() -> bool:
+    value = os.environ.get(_REGIONAL_COMPILE_DYNAMIC_ENV)
```
Comment (Gaohan123) [inline vllm_omni/diffusion/worker/diffusion_model_runner.py]: I don't suggest we use env variable. Please take it as input argument via cli or config

Comment (TaffyOfficial) [reply vllm_omni/diffusion/worker/diffusion_model_runner.py]: fix

**Comment by Gaohan123** [comment]
Please resolve conflicts

**Comment by TaffyOfficial** [comment]
> Please resolve conflicts

FIX

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks
